scan/src/utils/Route.re

```diff
@@ -39,87 +39,165 @@ type t =
+      | _ => NotFound
+      };
+    },
+  ),
+];
+
 let fromUrl = (url: ReasonReactRouter.url) =>
-  switch (url.path, url.hash) {
-  | (["data-sources"], _) => DataSourceHomePage
-  | (["data-source", dataSourceID], "code") =>
-    DataSourceIndexPage(dataSourceID |> int_of_string, DataSourceCode)
-  | (["data-source", dataSourceID], "requests") =>
-    DataSourceIndexPage(dataSourceID |> int_of_string, DataSourceRequests)
-  | (["data-source", dataSourceID], "revisions") =>
-    DataSourceIndexPage(dataSourceID |> int_of_string, DataSourceRevisions)
-  | (["data-source", dataSourceID], _) =>
-    DataSourceIndexPage(dataSourceID |> int_of_string, DataSourceExecute)
-  | (["oracle-scripts"], _) => OracleScriptHomePage
-  | (["oracle-script", oracleScriptID], "code") =>
-    OracleScriptIndexPage(oracleScriptID |> int_of_string, OracleScriptCode)
-  | (["oracle-script", oracleScriptID], "requests") =>
-    OracleScriptIndexPage(oracleScriptID |> int_of_string, OracleScriptRequests)
-  | (["oracle-script", oracleScriptID], "revisions") =>
-    OracleScriptIndexPage(oracleScriptID |> int_of_string, OracleScriptRevisions)
-  | (["oracle-script", oracleScriptID], _) =>
-    OracleScriptIndexPage(oracleScriptID |> int_of_string, OracleScriptExecute)
-  | (["txs"], _) => TxHomePage
-  | (["tx", txHash], _) => TxIndexPage(Hash.fromHex(txHash))
-  | (["validators"], _) => ValidatorHomePage
-  | (["blocks"], _) => BlockHomePage
-  | (["block", blockHeight], _) =>
-    let blockHeightIntOpt = blockHeight |> int_of_string_opt;
-    BlockIndexPage(blockHeightIntOpt->Belt_Option.getWithDefault(0));
-  | (["request", reqID], "proof") => RequestIndexPage(reqID |> int_of_string, RequestProof)
-  | (["request", reqID], _) => RequestIndexPage(reqID |> int_of_string, RequestReportStatus)
-  | (["account", address], "delegations") =>
-    AccountIndexPage(address |> Address.fromBech32, AccountDelegations)
-  | (["account", address], _) =>
-    AccountIndexPage(address |> Address.fromBech32, AccountTransactions)
-  | (["validator", address], "delegators") =>
-    ValidatorIndexPage(address |> Address.fromBech32, Delegators)
-  | (["validator", address], "reports") =>
-    ValidatorIndexPage(address |> Address.fromBech32, Reports)
-  | (["validator", address], _) =>
-    ValidatorIndexPage(address |> Address.fromBech32, ProposedBlocks)
-  | ([], _) => HomePage
-  | (_, _) => NotFound
+  switch (
+    regexes
+    ->Belt_List.keepMap(((regex, keysToRoute)) =>
+        url.path
+        ->Belt_List.get(0)
+        ->Belt_Option.getWithDefault("")
+        ->Js.Re.exec_(regex, _)
+        ->Belt_Option.map(result => (result, keysToRoute))
+      )
+    ->Belt.List.head
+  ) {
+  | Some((result, keysToRoute)) =>
+    keysToRoute(
+      result
+      ->Js.Re.captures
+      ->Belt_Array.keepMap(Js.toOption)
+      ->Belt_List.fromArray
+      ->Belt_List.concat(url.path->Belt_List.drop(1)->Belt_Option.getWithDefault([]))
```

Comment:
This is too much of a confusion. You are concating what you get from regex extractions with other parts of path. Don't do that.

Reply:
This is happen because "tx/abcd...".
Other paths don't have "/". So url.path only contain 1 element such as "D12", "bandvfsettg...".
But for tx, its path will contain 2 elements (["tx", "abcd..."]).

Reply:
No. let's split into separate variables